app/models/stage.rb

```diff
@@ -28,6 +28,8 @@ class Stage < ActiveRecord::Base
   attr_writer :command
   before_save :build_new_project_command
 
+  before_soft_delete :verify_not_part_of_pipeline
```

Comment:
if I move this to `plugins/pipelines/app/decorators/stage_decorator` I get an undefined method error for `before_soft_delete`.  I could move the `has_soft_deletion` call there too, but I'm not sure that makes sense. Maybe I'm trying to move it to the wrong spot?